### PR TITLE
ACCUMULO-3509: Make cleanup stateful to minimize blocking

### PR DESCRIPTION
**Pull request by phrocker**
By enabling state ( true/false) within the cleanup method, the change will avoid blocking
on a scan session being swept. if the session cleanup blocks because a ScanSession is
still being read, we may block until the ScanBatch returns for that ScanSession.

The change uses a simple semaphore ( purely because I like the word ) to attempt acquisition.
If that fails, we return false from the cleanup and reintroduce that Session back into
the queue to clean up.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 1]
```diff
@@ -226,6 +226,7 @@
   TSERV_MEM_MGMT("tserver.memory.manager", "org.apache.accumulo.server.tabletserver.LargestFirstMemoryManager", PropertyType.CLASSNAME,
       "An implementation of MemoryManger that accumulo will use."),
   TSERV_SESSION_MAXIDLE("tserver.session.idle.max", "1m", PropertyType.TIMEDURATION, "maximum idle time for a session"),
+  TSERV_UPDATE_SESSION_MAXIDLE("tserver.session.update.idle.max", "1m", PropertyType.TIMEDURATION, "maximum idle time for an update session"),
```
Comment (joshelser) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Could you make the new property description a complete sentence, please?

File: server/tserver/src/main/java/org/apache/accumulo/tserver/Tablet.java [hunk 2]
```diff
@@ -1761,33 +1763,41 @@ Scanner createScanner(Range range, int num, Set<Column> columns, Authorizations
...
-      }
+        try {
+          scannerSemaphore.acquire();
+        } catch (InterruptedException e) {
+          sawException = true;
+        }
 
-      try {
+        if (sawException)
+          throw new IllegalStateException("Tried to use scanner after exception occurred.");
```
Comment (joshelser) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/Tablet.java]: Isn't this exception message a little misleading now if we catch the InterruptedException above? Would it be better to throw an ISE from the catch block above with a better message? Or, do we not care what exception we saw?

File: server/tserver/src/main/java/org/apache/accumulo/tserver/Tablet.java [hunk 3]
```diff
@@ -1761,33 +1763,41 @@ Scanner createScanner(Range range, int num, Set<Column> columns, Authorizations
     private ScanDataSource isolatedDataSource;
     private boolean sawException = false;
     private boolean scanClosed = false;
+    private Semaphore scannerSemaphore;
```
Comment (joshelser) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/Tablet.java]: A javadoc comment describing why a semaphore is used would be nice when looking at this code in the future.

File: test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+    InstanceOperations ops = getConnector().instanceOperations();
+    sessionIdle = ops.getSystemConfiguration().get(Property.TSERV_SESSION_MAXIDLE.getKey());
+    ops.setProperty(Property.TSERV_SESSION_MAXIDLE.getKey(), "1s");
+    log.info("Waiting for existing session idle time to expire");
+    Thread.sleep(AccumuloConfiguration.getTimeInMillis(sessionIdle));
+    log.info("Finished waiting");
+  }
+
+  ExecutorService service = Executors.newFixedThreadPool(10);
```
Comment (joshelser) [inline test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java]: Create this inside the test case? Or in a `@BeforeClass`? Something that we can make sure it is cleaned up after the test completes.

File: test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+            slow.next();
+          }
+          return slow.hasNext();
+        }
+      });
+      callables.add(callable);
+    }
+
+    Thread.sleep(10000);
```
Comment (joshelser) [inline test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java]: Could you use some kind of construct (maybe a CountDownLatch) to guarantee at least one of your slow scans is running? Relying on a 10second pause is very likely to cause us test stability issues.

File: test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+      scanner2.setBatchSize(1);
+      Iterator<Entry<Key,Value>> iter = scanner2.iterator();
+      sinkKeyValues(iter);
+
+    }
+
+    int sessionsFound = 0;
+    // we have configured 1 tserver, so we can grab the one and only
+    String tserver = c.instanceOperations().getTabletServers().iterator().next();
```
Comment (joshelser) [inline test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java]: `Iterables.getOnlyElement(Iterable)` would be a little more concise than this and include better error checking.

File: test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+      if (tableName.equals(scan.getTable()) && scan.getSsiList().size() > 0) {
+        assertEquals("Not the expected iterator", 1, scan.getSsiList().size());
+        assertTrue("Not the expected iterator", scan.getSsiList().iterator().next().contains("SlowIterator"));
+        sessionsFound++;
+      }
+
+    }
+
+    assertEquals("Must have ten sessions to ensure 3509 is resolved", 10, sessionsFound);
```
Comment (joshelser) [inline test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java]: This is a little non-intuitive at a glance. Could you reword it to more clearly state why this equality check implies that ACCUMULO-3509 is resolved (instead of letting that up to the person reading a test failure)?

File: test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    for (Future<Boolean> callable : callables) {
+      callable.cancel(true);
+    }
+    service.shutdown();
+  }
+
+  private void sinkKeyValues(Iterator<Entry<Key,Value>> iter) throws Exception {
+    while (iter.hasNext()) {
```
Comment (joshelser) [inline test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java]: Could you do a quick summation of the values you read to make sure you got all of the expected key-value pairs?

**Comment by joshelser** [comment]
Looks pretty good, @phrocker. I left you some nit-picky suggestions which would improve the changes a bit and hopefully make sure your included integration test runs on any hardware (yay, included integration test!!).

File: test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java (hunk 5 above)
Comment (phrocker) [inline test/src/test/java/org/apache/accumulo/test/functional/SessionBlockVerifyIT.java]: You're right. this was a very lazy way of doing this. I will fix along with everything else you mentioned. Thanks for catching this.

**Comment by joshelser** [comment]
Thanks, @phrocker. Future-Josh thanks you too (hopefully your improvements will save me another round of "make our ITs stable".
